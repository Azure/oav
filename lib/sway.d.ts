declare module "sway" {
  interface Options {
    definition: any
    jsonRefs: { relativeBase: any }
    isPathCaseSensitive?: boolean
  }
  interface SwaggerApi {
    validate(): any
    getOperations(): any
    info: {
      version: string
      title: any
    }
  }
  function create(options: Options): Promise<SwaggerApi>
}