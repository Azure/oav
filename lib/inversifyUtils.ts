import { Container, interfaces } from "inversify";
import { setDefaultOpts } from "./swagger/loader";

export const TYPES = {
  opts: Symbol("InversifyTYPES.opts"),
};

export const inversifyGetInstance = <T, Opt = {}>(
  claz: interfaces.Newable<T>,
  opts: Opt &
    interfaces.ContainerOptions & {
      container?: Container;
    }
) => {
  if (opts.container === undefined) {
    setDefaultOpts(opts, {
      defaultScope: "Singleton",
      autoBindInjectable: true,
    } as any);
    opts.container = new Container(opts);
  }
  opts.container.bind(TYPES.opts).toConstantValue(opts);
  return opts.container.get(claz);
};
