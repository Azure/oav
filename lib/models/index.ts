/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */

export { LiveRequest } from './liveRequest'
export { LiveResponse } from './liveResponse'
export { RequestResponse } from './requestResponse'
export { OperationInfo } from './operationInfo'
export { LiveValidationError } from './liveValidationError'
export { ErrorResponse } from './errorResponse'
export { ErrorWrapper } from './errorWrapper'
export { RequestValidationResult } from './requestValidationResult'
export { ResponseValidationResult } from './responseValidationResult'
export { ValidationResult } from './validationResult'
export { PotentialOperationsResult } from './potentialOperationsResult'
