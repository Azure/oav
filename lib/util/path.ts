// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

import { flatMap } from "@ts-common/iterator"

export const splitPathAndReverse = (p: string | undefined) =>
  p === undefined ? undefined : Array.from(flatMap(p.split("/"), s => s.split("\\"))).reverse()

export const isSubPath = (
  mainPath: ReadonlyArray<string> | undefined,
  subPath: ReadonlyArray<string> | undefined,
) =>
  // return `true` if there are no subPath.
  subPath === undefined ||
  subPath.length === 0 ||
  // return `true` if `subPath` is a sub-path of `mainPath`.
  (mainPath !== undefined &&
    mainPath.length > subPath.length &&
    subPath.every((s, i) => mainPath[i] === s)
  )
