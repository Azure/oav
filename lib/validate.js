// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

'use strict';

var fs = require('fs'),
  path = require('path'),
  msrest = require('ms-rest'),
  msrestazure = require('ms-rest-azure'),
  ResourceManagementClient = require('azure-arm-resource').ResourceManagementClient,
  log = require('./util/logging'),
  utils = require('./util/utils'),
  Constants = require('./util/constants'),
  path = require('path'),
  util = require('util'),
  SpecValidator = require('./validators/specValidator'),
  WireFormatGenerator = require('./wireFormatGenerator'),
  XMsExampleExtractor = require('./xMsExampleExtractor'),
  SpecResolver = require('./validators/specResolver'),
  UmlGenerator = require('./umlGenerator');

exports = module.exports;

exports.finalValidationResult = { validityStatus: true };

exports.getDocumentsFromCompositeSwagger = function getDocumentsFromCompositeSwagger(compositeSpecPath) {
  let compositeSwagger;
  let finalDocs = [];
  return utils.parseJson(compositeSpecPath).then(function (result) {
    compositeSwagger = result;
    if (!(compositeSwagger.documents && Array.isArray(compositeSwagger.documents) && compositeSwagger.documents.length > 0)) {
      throw new Error(`CompositeSwagger - ${compositeSpecPath} must contain a documents property and it must be of type array and it must be a non empty array.`);
    }
    let docs = compositeSwagger.documents;
    let basePath = path.dirname(compositeSpecPath);
    for (let i = 0; i < docs.length; i++) {
      if (docs[i].startsWith('.')) {
        docs[i] = docs[i].substring(1);
      }
      let individualPath = '';
      if (docs[i].startsWith('http')) {
        individualPath = docs[i];
      } else {
        individualPath = basePath + docs[i];
      }
      finalDocs.push(individualPath);
    }
    return finalDocs;
  }).catch(function (err) {
    log.error(err);
    return Promise.reject(err);
  });
};

exports.validateSpec = function validateSpec(specPath, options) {
  if (!options) options = {};
  log.consoleLogLevel = options.consoleLogLevel || log.consoleLogLevel;
  log.filepath = options.logFilepath || log.filepath;
  // As a part of resolving discriminators we replace all the parent references
  // with a oneof array containing references to the parent and its children.
  // This breaks the swagger specification 2.0 schema since oneOf is not supported.
  // Hence we disable it since it is not required for semantic check.

  options.shouldResolveDiscriminator = false;
  // parameters in 'x-ms-parameterized-host' extension need not be resolved for semantic
  // validation as that would not match the path parameters defined in the path template
  // and cause the semantic validation to fail.
  options.shouldResolveParameterizedHost = false;
  let validator = new SpecValidator(specPath, null, options);
  exports.finalValidationResult[specPath] = validator.specValidationResult;
  return validator.initialize().then(function () {
    log.info(`Semantically validating  ${specPath}:\n`);
    return validator.validateSpec().then(function (result) {
      exports.updateEndResultOfSingleValidation(validator);
      exports.logDetailedInfo(validator);
      return Promise.resolve(validator.specValidationResult);
    });
  }).catch(function (err) {
    log.error(err);
    return Promise.reject(err);
  });
};

exports.validateCompositeSpec = function validateCompositeSpec(compositeSpecPath, options) {
  if (!options) options = {};
  log.consoleLogLevel = options.consoleLogLevel || log.consoleLogLevel;
  log.filepath = options.logFilepath || log.filepath;
  return exports.getDocumentsFromCompositeSwagger(compositeSpecPath).then(function (docs) {
    options.consoleLogLevel = log.consoleLogLevel;
    options.logFilepath = log.filepath;
    let promiseFactories = docs.map(function (doc) {
      return function () { return exports.validateSpec(doc, options) };
    });
    return utils.executePromisesSequentially(promiseFactories);
  }).catch(function (err) {
    log.error(err);
    return Promise.reject(err);
  });
};

exports.validateExamples = function validateExamples(specPath, operationIds, options) {
  if (!options) options = {};
  log.consoleLogLevel = options.consoleLogLevel || log.consoleLogLevel;
  log.filepath = options.logFilepath || log.filepath;
  let validator = new SpecValidator(specPath, null, options);
  exports.finalValidationResult[specPath] = validator.specValidationResult;
  return validator.initialize().then(function () {
    log.info(`Validating "examples" and "x-ms-examples" in  ${specPath}:\n`);
    validator.validateOperations(operationIds);
    exports.updateEndResultOfSingleValidation(validator);
    exports.logDetailedInfo(validator);
    return Promise.resolve(validator.specValidationResult);
  }).catch(function (err) {
    log.error(err);
    return Promise.reject(err);
  });
};

exports.validateExamplesInCompositeSpec = function validateExamplesInCompositeSpec(compositeSpecPath, options) {
  if (!options) options = {};
  log.consoleLogLevel = options.consoleLogLevel || log.consoleLogLevel;
  log.filepath = options.logFilepath || log.filepath;
  return exports.getDocumentsFromCompositeSwagger(compositeSpecPath).then(function (docs) {
    options.consoleLogLevel = log.consoleLogLevel;
    options.logFilepath = log.filepath;
    let promiseFactories = docs.map(function (doc) {
      return function () { return exports.validateExamples(doc, options); }
    });
    return utils.executePromisesSequentially(promiseFactories);
  }).catch(function (err) {
    log.error(err);
    return Promise.reject(err);
  });
};

exports.resolveSpec = function resolveSpec(specPath, outputDir, options) {
  if (!options) options = {};
  log.consoleLogLevel = options.consoleLogLevel || log.consoleLogLevel;
  log.filepath = options.logFilepath || log.filepath;
  let specFileName = path.basename(specPath);
  let resolver;
  return utils.parseJson(specPath).then((result) => {
    resolver = new SpecResolver(specPath, result, options);
    return resolver.resolve();
  }).then(() => {
    let resolvedSwagger = JSON.stringify(resolver.specInJson, null, 2);
    if (outputDir !== './' && !fs.existsSync(outputDir)) {
      fs.mkdirSync(outputDir);
    }
    let outputFilepath = `${path.join(outputDir, specFileName)}`;
    fs.writeFileSync(`${path.join(outputDir, specFileName)}`, resolvedSwagger, { encoding: 'utf8' });
    console.log(`Saved the resolved spec at "${outputFilepath}".`)
    return Promise.resolve();
  }).catch((err) => {
    log.error(err);
    return Promise.reject(err);
  });
};

exports.resolveCompositeSpec = function resolveCompositeSpec(specPath, outputDir, options) {
  if (!options) options = {};
  log.consoleLogLevel = options.consoleLogLevel || log.consoleLogLevel;
  log.filepath = options.logFilepath || log.filepath;
  return exports.getDocumentsFromCompositeSwagger(compositeSpecPath).then(function (docs) {
    options.consoleLogLevel = log.consoleLogLevel;
    options.logFilepath = log.filepath;
    let promiseFactories = docs.map(function (doc) {
      return function () { return exports.resolveSpec(doc, outputDir, options); }
    });
    return utils.executePromisesSequentially(promiseFactories);
  }).catch(function (err) {
    log.error(err);
    return Promise.reject(err);
  });
};


exports.generateWireFormat = function generateWireFormat(specPath, outDir, emitYaml, operationIds, options) {
  if (!options) options = {};
  log.consoleLogLevel = options.consoleLogLevel || log.consoleLogLevel;
  log.filepath = options.logFilepath || log.filepath;
  let wfGenerator = new WireFormatGenerator(specPath, null, outDir, emitYaml);
  return wfGenerator.initialize().then(function () {
    log.info(`Generating wire format request and responses for swagger spec: "${specPath}":\n`);
    wfGenerator.processOperations(operationIds);
    return Promise.resolve();
  }).catch(function (err) {
    log.error(err);
    return Promise.reject(err);
  });
};

exports.generateWireFormatInCompositeSpec = function generateWireFormatInCompositeSpec(compositeSpecPath, outDir, emitYaml, options) {
  if (!options) options = {};
  log.consoleLogLevel = options.consoleLogLevel || log.consoleLogLevel;
  log.filepath = options.logFilepath || log.filepath;
  return exports.getDocumentsFromCompositeSwagger(compositeSpecPath).then(function (docs) {
    options.consoleLogLevel = log.consoleLogLevel;
    options.logFilepath = log.filepath;
    let promiseFactories = docs.map(function (doc) {
      return function () { return exports.generateWireFormat(doc, outDir, emitYaml, null, options); }
    });
    return utils.executePromisesSequentially(promiseFactories);
  }).catch(function (err) {
    log.error(err);
    return Promise.reject(err);
  });
};

exports.generateUml = function generateUml(specPath, outputDir, options) {
  if (!options) options = {};
  log.consoleLogLevel = options.consoleLogLevel || log.consoleLogLevel;
  log.filepath = options.logFilepath || log.filepath;
  let specFileName = path.basename(specPath);
  let resolver;
  let resolverOptions = {};
  resolverOptions.shouldResolveRelativePaths = true;
  resolverOptions.shouldResolveXmsExamples = false;
  resolverOptions.shouldResolveAllOf = false;
  resolverOptions.shouldSetAdditionalPropertiesFalse = false;
  resolverOptions.shouldResolvePureObjects = false;
  resolverOptions.shouldResolveDiscriminator = false;
  resolverOptions.shouldResolveParameterizedHost = false;
  resolverOptions.shouldResolveNullableTypes = false;
  return utils.parseJson(specPath).then((result) => {
    resolver = new SpecResolver(specPath, result, resolverOptions);
    return resolver.resolve();
  }).then(() => {
    let umlGenerator = new UmlGenerator(resolver.specInJson, options);
    return umlGenerator.generateDiagramFromGraph();
  }).then((svgGraph) => {
    if (outputDir !== './' && !fs.existsSync(outputDir)) {
      fs.mkdirSync(outputDir);
    }
    let svgFile = specFileName.replace(path.extname(specFileName), '.svg');
    let outputFilepath = `${path.join(outputDir, svgFile)}`;
    fs.writeFileSync(`${path.join(outputDir, svgFile)}`, svgGraph, { encoding: 'utf8' });
    console.log(`Saved the uml at "${outputFilepath}". Please open the file in a browser.`);
    return Promise.resolve();
  }).catch((err) => {
    log.error(err);
    return Promise.reject(err);
  });
};

exports.updateEndResultOfSingleValidation = function updateEndResultOfSingleValidation(validator) {
  if (validator.specValidationResult.validityStatus) {
    if (!(log.consoleLogLevel === 'json' || log.consoleLogLevel === 'off')) {
      let consoleLevel = log.consoleLogLevel;
      log.consoleLogLevel = 'info';
      log.info('No Errors were found.');
      log.consoleLogLevel = consoleLevel;
    }
  }
  if (!validator.specValidationResult.validityStatus) {
    process.exitCode = 1;
    exports.finalValidationResult.validityStatus = validator.specValidationResult.validityStatus;
  }
  return;
};

exports.logDetailedInfo = function logDetailedInfo(validator) {
  if (log.consoleLogLevel === 'json') {
    console.dir(validator.specValidationResult, { depth: null, colors: true });
  }
  log.silly('############################');
  log.silly(validator.specValidationResult);
  log.silly('----------------------------');
};

exports.extractXMsExamples = function extractXMsExamples(specPath, recordings, options) {
  if (!options) options = {};
  log.consoleLogLevel = options.consoleLogLevel || log.consoleLogLevel;
  log.filepath = options.logFilepath || log.filepath;
  let xMsExampleExtractor = new XMsExampleExtractor(specPath, recordings, options);
  return xMsExampleExtractor.extract();
};
