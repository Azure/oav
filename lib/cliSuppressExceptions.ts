import { log } from './util/logging';

// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

export const cliSuppressExceptions = async (f: () => Promise<void>): Promise<void> => {
  try {
    await f()
  } catch (err) {
    const message = `fatal error: ${JSON.stringify(err)}`
    log.error(message)
    // tslint:disable-next-line:no-console
    console.error(message)
    process.exitCode = 1
  }
}
