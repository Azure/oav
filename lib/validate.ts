// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

import { Suppression } from "@azure/openapi-markdown"
import * as jsonParser from "@ts-common/json-parser"
import { StringMap } from "@ts-common/string-map"
import * as fs from "fs"
import jsYaml from "js-yaml"
import * as path from "path"

import * as umlGeneratorLib from "./umlGenerator"
import { getErrorsFromModelValidation } from "./util/getErrorsFromModelValidation"
import * as jsonUtils from "./util/jsonUtils"
import { log } from "./util/logging"
import { ModelValidationError } from "./util/modelValidationError"
import * as utils from "./util/utils"
import { NodeError } from "./util/validationError"
import { ModelValidator } from "./validators/modelValidator"
import { SemanticValidator } from "./validators/semanticValidator"
import * as specResolver from "./validators/specResolver"
import {
  CommonValidationResult,
  SpecValidationResult,
  SpecValidator
} from "./validators/specValidator"
import { getSuppressions } from "./validators/suppressions"
import { WireFormatGenerator } from "./wireFormatGenerator"
import { XMsExampleExtractor } from "./xMsExampleExtractor"

export interface Options extends specResolver.Options, umlGeneratorLib.Options {
  consoleLogLevel?: unknown
  logFilepath?: unknown
  pretty?: boolean
}

export async function getDocumentsFromCompositeSwagger(
  suppression: Suppression | undefined,
  compositeSpecPath: string,
  reportError: jsonParser.ReportError
): Promise<string[]> {
  try {
    const compositeSwagger = await jsonUtils.parseJson(suppression, compositeSpecPath, reportError)
    if (
      !(
        compositeSwagger.documents &&
        Array.isArray(compositeSwagger.documents) &&
        compositeSwagger.documents.length > 0
      )
    ) {
      throw new Error(
        `CompositeSwagger - ${compositeSpecPath} must contain a documents property and it must ` +
          `be of type array and it must be a non empty array.`
      )
    }
    const docs = compositeSwagger.documents
    const basePath = path.dirname(compositeSpecPath)
    const finalDocs: string[] = []
    for (let i = 0; i < docs.length; i++) {
      if (docs[i].startsWith(".")) {
        docs[i] = docs[i].substring(1)
      }
      const individualPath = docs[i].startsWith("http") ? docs[i] : basePath + docs[i]
      finalDocs.push(individualPath)
    }
    return finalDocs
  } catch (err) {
    log.error(err)
    throw err
  }
}

const vsoLogIssueWrapper = (issueType: string, message: string) => {
  return issueType === "error" || issueType === "warning"
    ? `##vso[task.logissue type=${issueType}]${message}`
    : `##vso[task.logissue type=error]${message}`
}

async function validate<T>(
  options: Options | undefined,
  func: (options: Options) => Promise<T>
): Promise<T> {
  if (!options) {
    options = {}
  }
  log.consoleLogLevel = options.consoleLogLevel || log.consoleLogLevel
  log.filepath = options.logFilepath || log.filepath
  if (options.pretty) {
    log.consoleLogLevel = "off"
  }
  return func(options)
}

type ErrorType = "error" | "warning"

const prettyPrint = <T extends NodeError<T>>(
  errors: ReadonlyArray<T> | undefined,
  errorType: ErrorType
) => {
  if (errors !== undefined) {
    for (const error of errors) {
      const yaml = jsYaml.dump(error)
      if (process.env["Agent.Id"]) {
        /* tslint:disable-next-line:no-console no-string-literal */
        console.error(vsoLogIssueWrapper(errorType, errorType))
        /* tslint:disable-next-line:no-console no-string-literal */
        console.error(vsoLogIssueWrapper(errorType, yaml))
      } else {
        /* tslint:disable-next-line:no-console no-string-literal */
        console.error("\x1b[31m", errorType, ":", "\x1b[0m")
        /* tslint:disable-next-line:no-console no-string-literal */
        console.error(yaml)
      }
    }
  }
}

export async function validateSpec(
  specPath: string,
  options: Options | undefined
): Promise<SpecValidationResult> {
  return validate(options, async o => {
    // As a part of resolving discriminators we replace all the parent references
    // with a oneOf array containing references to the parent and its children.
    // This breaks the swagger specification 2.0 schema since oneOf is not supported.
    // Hence we disable it since it is not required for semantic check.

    o.shouldResolveDiscriminator = false
    // parameters in 'x-ms-parameterized-host' extension need not be resolved for semantic
    // validation as that would not match the path parameters defined in the path template
    // and cause the semantic validation to fail.
    o.shouldResolveParameterizedHost = false

    // We shouldn't be resolving nullable types for semantic validation as we'll replace nodes
    // with oneOf arrays which are not semantically valid in swagger 2.0 schema.
    o.shouldResolveNullableTypes = false

    const validator = new SemanticValidator(specPath, null, o)
    const suppression = await getSuppressions(specPath)
    await validator.initialize(suppression)
    log.info(`Semantically validating  ${specPath}:\n`)
    const validationResults = await validator.validateSpec(specPath, suppression)
    updateEndResultOfSingleValidation(validator)
    logDetailedInfo(validator)
    if (o.pretty) {
      /* tslint:disable-next-line:no-console no-string-literal */
      console.log(vsoLogIssueWrapper("error", `Semantically validating  ${specPath}:\n`))
      const resolveSpecError = validator.specValidationResult.resolveSpec
      if (resolveSpecError !== undefined) {
        prettyPrint([resolveSpecError], "error")
      }
      prettyPrint(validationResults.errors, "error")
      prettyPrint(validationResults.warnings, "warning")
    }
    return validator.specValidationResult
  })
}

export async function validateCompositeSpec(
  compositeSpecPath: string,
  options: Options
): Promise<ReadonlyArray<SpecValidationResult>> {
  return validate(options, async o => {
    const suppression = await getSuppressions(compositeSpecPath)
    const docs = await getDocumentsFromCompositeSwagger(
      suppression,
      compositeSpecPath,
      jsonParser.defaultErrorReport
    )
    o.consoleLogLevel = log.consoleLogLevel
    o.logFilepath = log.filepath
    const promiseFactories = docs.map(doc => async () => validateSpec(doc, o))
    return utils.executePromisesSequentially(promiseFactories)
  })
}

export async function validateExamples(
  specPath: string,
  operationIds: string | undefined,
  options?: Options
): Promise<ReadonlyArray<ModelValidationError>> {
  return validate(options, async o => {
    const validator = new ModelValidator(specPath, null, o)
    await validator.initialize()
    log.info(`Validating "examples" and "x-ms-examples" in  ${specPath}:\n`)
    await validator.validateOperations(operationIds)
    updateEndResultOfSingleValidation(validator)
    logDetailedInfo(validator)
    const errors = getErrorsFromModelValidation(validator.specValidationResult)
    if (o.pretty) {
      /* tslint:disable-next-line:no-console no-string-literal */
      console.log(
        vsoLogIssueWrapper("info", `Validating "examples" and "x-ms-examples" in  ${specPath}:\n`)
      )

      if (errors.length > 0) {
        prettyPrint(errors, "error")
      }
    }
    return errors
  })
}

export async function validateExamplesInCompositeSpec(
  compositeSpecPath: string,
  options: Options
): Promise<ReadonlyArray<ReadonlyArray<ModelValidationError>>> {
  return validate(options, async o => {
    o.consoleLogLevel = log.consoleLogLevel
    o.logFilepath = log.filepath
    const suppression = await getSuppressions(compositeSpecPath)
    const docs = await getDocumentsFromCompositeSwagger(
      suppression,
      compositeSpecPath,
      jsonParser.defaultErrorReport
    )
    const promiseFactories = docs.map(doc => async () => validateExamples(doc, undefined, o))
    return utils.executePromisesSequentially(promiseFactories)
  })
}

export async function resolveSpec(
  specPath: string,
  outputDir: string,
  options: Options,
  reportError: jsonParser.ReportError
): Promise<void> {
  if (!options) {
    options = {}
  }
  log.consoleLogLevel = options.consoleLogLevel || log.consoleLogLevel
  log.filepath = options.logFilepath || log.filepath
  const specFileName = path.basename(specPath)
  try {
    const suppression = await getSuppressions(specPath)
    const result = await jsonUtils.parseJson(suppression, specPath, reportError)
    const resolver = new specResolver.SpecResolver(specPath, result, options, reportError)
    await resolver.resolve(suppression)
    const resolvedSwagger = JSON.stringify(resolver.specInJson, null, 2)
    if (outputDir !== "./" && !fs.existsSync(outputDir)) {
      fs.mkdirSync(outputDir)
    }
    const outputFilepath = `${path.join(outputDir, specFileName)}`
    fs.writeFileSync(`${path.join(outputDir, specFileName)}`, resolvedSwagger, {
      encoding: "utf8"
    })
    /* tslint:disable-next-line */
    console.log(`Saved the resolved spec at "${outputFilepath}".`)
  } catch (err) {
    log.error(err)
    throw err
  }
}

export async function resolveCompositeSpec(
  specPath: string,
  outputDir: string,
  options: Options
): Promise<void> {
  if (!options) {
    options = {}
  }
  log.consoleLogLevel = options.consoleLogLevel || log.consoleLogLevel
  log.filepath = options.logFilepath || log.filepath
  try {
    const suppression = await getSuppressions(specPath)
    const docs = await getDocumentsFromCompositeSwagger(
      suppression,
      specPath,
      jsonParser.defaultErrorReport
    )
    options.consoleLogLevel = log.consoleLogLevel
    options.logFilepath = log.filepath
    const promiseFactories = docs.map(doc => async () =>
      resolveSpec(doc, outputDir, options, jsonParser.defaultErrorReport)
    )
    await utils.executePromisesSequentially(promiseFactories)
  } catch (err) {
    log.error(err)
    throw err
  }
}

export async function generateWireFormat(
  specPath: string,
  outDir: string,
  emitYaml: unknown,
  operationIds: string | null,
  options: Options
): Promise<void> {
  if (!options) {
    options = {}
  }
  log.consoleLogLevel = options.consoleLogLevel || log.consoleLogLevel
  log.filepath = options.logFilepath || log.filepath
  const wfGenerator = new WireFormatGenerator(specPath, null, outDir, emitYaml)
  try {
    await wfGenerator.initialize()
    log.info(`Generating wire format request and responses for swagger spec: "${specPath}":\n`)
    wfGenerator.processOperations(operationIds)
  } catch (err) {
    log.error(err)
    throw err
  }
}

export async function generateWireFormatInCompositeSpec(
  compositeSpecPath: string,
  outDir: string,
  emitYaml: unknown,
  options: Options
): Promise<void> {
  if (!options) {
    options = {}
  }
  log.consoleLogLevel = options.consoleLogLevel || log.consoleLogLevel
  log.filepath = options.logFilepath || log.filepath
  try {
    const suppression = await getSuppressions(compositeSpecPath)
    const docs = await getDocumentsFromCompositeSwagger(
      suppression,
      compositeSpecPath,
      jsonParser.defaultErrorReport
    )
    options.consoleLogLevel = log.consoleLogLevel
    options.logFilepath = log.filepath
    const promiseFactories = docs.map(doc => async () =>
      generateWireFormat(doc, outDir, emitYaml, null, options)
    )
    await utils.executePromisesSequentially(promiseFactories)
  } catch (err) {
    log.error(err)
    throw err
  }
}

export async function generateUml(
  specPath: string,
  outputDir: string,
  options?: Options
): Promise<void> {
  if (!options) {
    options = {}
  }
  log.consoleLogLevel = options.consoleLogLevel || log.consoleLogLevel
  log.filepath = options.logFilepath || log.filepath
  const specFileName = path.basename(specPath)
  const resolverOptions = {
    shouldResolveRelativePaths: true,
    shouldResolveXmsExamples: false,
    shouldResolveAllOf: false,
    shouldSetAdditionalPropertiesFalse: false,
    shouldResolvePureObjects: false,
    shouldResolveDiscriminator: false,
    shouldResolveParameterizedHost: false,
    shouldResolveNullableTypes: false
  }
  try {
    const suppression = await getSuppressions(specPath)
    const result = await jsonUtils.parseJson(suppression, specPath, jsonParser.defaultErrorReport)
    const resolver = new specResolver.SpecResolver(
      specPath,
      result,
      resolverOptions,
      jsonParser.defaultErrorReport
    )
    const umlGenerator = new umlGeneratorLib.UmlGenerator(resolver.specInJson, options)
    const svgGraph = await umlGenerator.generateDiagramFromGraph()
    if (outputDir !== "./" && !fs.existsSync(outputDir)) {
      fs.mkdirSync(outputDir)
    }
    const svgFile = specFileName.replace(path.extname(specFileName), ".svg")
    const outputFilepath = `${path.join(outputDir, svgFile)}`
    fs.writeFileSync(`${path.join(outputDir, svgFile)}`, svgGraph, {
      encoding: "utf8"
    })
    /* tslint:disable-next-line */
    console.log(`Saved the uml at "${outputFilepath}". Please open the file in a browser.`)
  } catch (err) {
    log.error(err)
    throw err
  }
}

export function updateEndResultOfSingleValidation<T extends CommonValidationResult>(
  validator: SpecValidator<T>
): void {
  if (validator.specValidationResult.validityStatus) {
    if (!(log.consoleLogLevel === "json" || log.consoleLogLevel === "off")) {
      log.info("No Errors were found.")
    }
  }
  if (!validator.specValidationResult.validityStatus) {
    process.exitCode = 1
  }
}

export function logDetailedInfo<T extends CommonValidationResult>(
  validator: SpecValidator<T>
): void {
  if (log.consoleLogLevel === "json") {
    /* tslint:disable-next-line */
    console.dir(validator.specValidationResult, { depth: null, colors: true })
  }
  log.silly("############################")
  log.silly(validator.specValidationResult.toString())
  log.silly("----------------------------")
}

export async function extractXMsExamples(
  specPath: string,
  recordings: string,
  options: Options
): Promise<StringMap<unknown>> {
  if (!options) {
    options = {}
  }
  log.consoleLogLevel = options.consoleLogLevel || log.consoleLogLevel
  log.filepath = options.logFilepath || log.filepath
  const xMsExampleExtractor = new XMsExampleExtractor(specPath, recordings, options)
  return xMsExampleExtractor.extract()
}
