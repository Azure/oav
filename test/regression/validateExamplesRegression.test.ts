import glob = require("glob")
import * as _ from "lodash"
import md5File = require("md5-file")
import * as path from "path"

import * as validate from "../../lib/validate"

describe("validateExamples should not regress for file", () => {
  const specPaths = glob
    .sync(path.join(__dirname, "azure-rest-api-specs/specification/**/*.json"))
    .filter(p => !p.includes("examples"))

  const versionRegex = /[0-9]+[0-9-]+[^\/]*/
  const rpRegex = /(?:\/)Microsoft.[^\/]*/
  const latestForEachRp = (_.chain(specPaths)
    .filter(p => p.includes("stable")) // only look at stable specs
    .filter(p => p.match(rpRegex))
    .groupBy((p: string) => p.match(rpRegex)![0]) // group paths by rp
    .entries()
    .map((entry: [string, string[]]) =>
      // for each rp get latest version
      _.chain(entry[1])
        .filter(p => p!.match(versionRegex))
        .groupBy((p: string) => p!.match(versionRegex)![0])
        .entries()
        .sortBy(([version]: [string, string[]]) => version)
        .map((e: [string, string[]]) => e[1])
        .last()
        .value()
    )
    .flatten()
    .value() as unknown) as string[]

  test.each(specPaths.map(path => [path, latestForEachRp.includes(path) ? "@latestVersion" : ""]))(
    "'%s' %s",
    async file => {
      try {
        const hash = await new Promise(resolve =>
          md5File(file, (e, h) => {
            if (e) {
              throw e
            }
            resolve(h)
          })
        )
        expect(hash).toMatchSnapshot("input file hash")

        const result = await validate.validateExamples(file, undefined, {
          pretty: true
        })
        expect(result).toMatchSnapshot("returned results")
      } catch (e) {
        expect(e).toMatchSnapshot("thrown exception")
      }
    },
    999999
  )
})
