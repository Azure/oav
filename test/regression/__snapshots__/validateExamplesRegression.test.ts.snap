// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@regression validateExamples should not regress for file '/Users/billytrend/Repositories/oav/test/regression/azure-rest-api-specs/specification/addons/resource-manager/Microsoft.Addons/preview/2017-05-15/Addons.json': input file hash 1`] = `null`;

exports[`@regression validateExamples should not regress for file '/Users/billytrend/Repositories/oav/test/regression/azure-rest-api-specs/specification/addons/resource-manager/Microsoft.Addons/preview/2017-05-15/Addons.json': returned results 1`] = `Array []`;

exports[`@regression validateExamples should not regress for file '/Users/billytrend/Repositories/oav/test/regression/azure-rest-api-specs/specification/addons/resource-manager/Microsoft.Addons/preview/2017-05-15/examples/CanonicalSupportPlanTypes_Get.json': input file hash 1`] = `null`;

exports[`@regression validateExamples should not regress for file '/Users/billytrend/Repositories/oav/test/regression/azure-rest-api-specs/specification/addons/resource-manager/Microsoft.Addons/preview/2017-05-15/examples/CanonicalSupportPlanTypes_Get.json': returned results 1`] = `Array []`;

exports[`@regression validateExamples should not regress for file '/Users/billytrend/Repositories/oav/test/regression/azure-rest-api-specs/specification/addons/resource-manager/Microsoft.Addons/preview/2017-05-15/examples/Operations_List.json': input file hash 1`] = `null`;

exports[`@regression validateExamples should not regress for file '/Users/billytrend/Repositories/oav/test/regression/azure-rest-api-specs/specification/addons/resource-manager/Microsoft.Addons/preview/2017-05-15/examples/Operations_List.json': returned results 1`] = `Array []`;

exports[`@regression validateExamples should not regress for file '/Users/billytrend/Repositories/oav/test/regression/azure-rest-api-specs/specification/addons/resource-manager/Microsoft.Addons/preview/2017-05-15/examples/SupportPlanTypes_CreateOrUpdate.json': input file hash 1`] = `null`;

exports[`@regression validateExamples should not regress for file '/Users/billytrend/Repositories/oav/test/regression/azure-rest-api-specs/specification/addons/resource-manager/Microsoft.Addons/preview/2017-05-15/examples/SupportPlanTypes_CreateOrUpdate.json': returned results 1`] = `Array []`;

exports[`@regression validateExamples should not regress for file '/Users/billytrend/Repositories/oav/test/regression/azure-rest-api-specs/specification/addons/resource-manager/Microsoft.Addons/preview/2017-05-15/examples/SupportPlanTypes_Delete.json': input file hash 1`] = `null`;

exports[`@regression validateExamples should not regress for file '/Users/billytrend/Repositories/oav/test/regression/azure-rest-api-specs/specification/addons/resource-manager/Microsoft.Addons/preview/2017-05-15/examples/SupportPlanTypes_Delete.json': returned results 1`] = `Array []`;
